target/linux/ramips/mt7621/base-files/etc/board.d/02_network

```diff
@@ -218,6 +219,14 @@ ramips_setup_macs()
 		lan_mac=$(mtd_get_mac_ascii factory lanmac)
 		wan_mac=$(mtd_get_mac_ascii factory wanmac)
 		;;
+	dna,valokuitu-plus-ex400)
+		lan_mac=$(mtd_get_mac_ascii_ubi "ubi" "env1" "ethaddr")
+		[ -z "$lan_mac" ] && lan_mac=$(mtd_get_mac_ascii_ubi "ubi" "env2" "ethaddr")
+		lan_mac=$(macaddr_canonicalize "$lan_mac")
+		wan_mac=$(macaddr_add "$lan_mac" 1)
+		label_mac="$lan_mac"
+		ucidef_set_network_device_mac "eth0" $label_mac
```

Comment:
I think you can use fw_printenv given that you already added support for it in uboot-envtools

Reply:
Hmm? NVMEM can’t be used here?

Reply:
Last time I checked, NVMEM on UBI has not been supported. There are are some patches from @dangowrt at https://github.com/dangowrt/linux/commits/for-nvmem-next, but last time I checked, they had been merged neither upstream nor as a patch in OpenWrt.

Reply:
Sorry, I didn't notice these comments earlier.
@Rondom using the fw_printenv was my initial attempt but the uboot-envtools has not been initialised when this script is being run.

@neheb I havent found any place where the mac would be in a flash but the u-boot env.